File: 053/MaximumSubArray.scala
object MaximumSubArray {
  def maxSubArray(nums: Array[Int]): Int = {
    var res = Int.MinValue
    nums.foldLeft(0) { (acc, num) =>
      var t = if (acc < 0) num else acc + num
      res = res max t
      t
    }
    res
  }

  def main(args: Array[String]): Unit = {
    var nums = Array(-2,1,-3,4,-1,2,1,-5,4)
    println(maxSubArray(nums))
  }

}

File: 053/MaximumSubArray.kt
class MaximumSubarray {
    fun maxSubArray(nums: IntArray): Int {
        val len = nums.size
        var max = nums[0]
        var dp = nums[0]
        for (i in 1 until len) {
            dp = Math.max(dp + nums[i], nums[i])
            max = Math.max(max, dp)
        }
        return max
    }
}

fun main(args: Array<String>) {
    println(MaximumSubarray().maxSubArray(intArrayOf(-2, 1, -3, 4, -1, 2, 1, -5, 4)))
}